src/CachingGenerator.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bigwhoop\SentenceBreaker;
+
+
+use Generator;
+
+class CachingGenerator implements \Iterator
+{
+    /** @var Generator */
+    private $generator;
+
+    /** @var array */
+    private $cache = [];
+
+    public function __construct(Generator $generator)
+    {
+        $this->generator = $generator;
+        $this->addCurrentToCache();
+    }
+
+    public function current()
+    {
+        return current($this->cache);
+    }
+
+    public function next(): void
+    {
+        if ($this->generator->key() === key($this->cache)) {
+            $this->generator->next();
+            $this->addCurrentToCache();
+        }
+        next($this->cache);
+    }
+
+    public function key(): ?int
+    {
+        return key($this->cache);
```

Comment:
TODO:
get rid of `key`/`reset`/`next` calls an replace implicit array-pointer with explicit `$currentPosition` property.